statsmodels/tsa/vector_ar/var_model.py

```diff
@@ -1342,39 +1678,188 @@ def test_causality(self, equation, variables, kind='f', signif=0.05,
+            * "pvalue" : float
+                The p-value corresponding to the test statistic.
+            * "df" : float
+                The degrees of freedom of the \Chi^2-distribution.
+            * "conclusion" : str {"reject", "fail to reject"}
+                 Whether H0 can be rejected or not.
+            * "signif" : float
+
+        Notes
+        -----
+        This method is not returning the same result as JMulTi. This is because
+        the test is based on a VAR(k_ar) model in statsmodels (in accordance to
+        pp. 104, 320-321 in [1]_) whereas JMulTi seems to be using a
+        VAR(k_ar+1) model.
+
+        References
+        ----------
+        .. [1] Lutkepohl, H. 2005. *New Introduction to Multiple Time Series Analysis*. Springer.
+        """
+        if not (0 < signif < 1):
+            raise ValueError("signif has to be between 0 and 1")
+
+        allowed_types = (string_types, int)
+        if isinstance(causing, allowed_types):
+            causing = [causing]
+        if not all(isinstance(c, allowed_types) for c in causing):
+            raise TypeError("causing has to be of type string or int (or a " +
+                            "a sequence of these types).")
+        causing = [self.names[c] if type(c) == int else c for c in causing]
+        causing_ind = [util.get_index(self.names, c) for c in causing]
+
+        caused_ind = [i for i in range(self.neqs) if i not in causing_ind]
+        caused = [self.names[c] for c in caused_ind]
+
+        # Note: JMulTi seems to be using k_ar+1 instead of k_ar
+        k, t, p = self.neqs, self.nobs, self.k_ar
+
+        num_restr = len(causing) * len(caused)  # called N in Lutkepohl
+
+        sigma_u = self.sigma_u
+        vech_sigma_u = util.vech(sigma_u)
+        sig_mask = np.zeros(sigma_u.shape)
+        # set =1 twice to ensure, that all the ones needed are below the main
+        # diagonal:
+        sig_mask[causing_ind, caused_ind] = 1
+        sig_mask[caused_ind, causing_ind] = 1
+        vech_sig_mask = util.vech(sig_mask)
+        inds = np.nonzero(vech_sig_mask)[0]
+
+        # Make restriction matrix
+        C = np.zeros((num_restr, len(vech_sigma_u)), dtype=float)
+        for row in range(num_restr):
+            C[row, inds[row]] = 1
+        Cs = np.dot(C, vech_sigma_u)
+        d = np.linalg.pinv(duplication_matrix(k))
+        Cd = np.dot(C, d)
+        middle = L.inv(chain_dot(Cd, np.kron(sigma_u, sigma_u), Cd.T)) / 2
+
+        wald_statistic = t * chain_dot(Cs.T, middle, Cs)
+        df = num_restr
+        dist = stats.chi2(df)
+
+        pvalue = dist.sf(wald_statistic)
+        crit_value = dist.ppf(1 - signif)
+
+        return CausalityTestResults(causing, caused, wald_statistic,
+                                    crit_value, pvalue, df, signif,
+                                    test="inst", method="wald")
+
+    def test_whiteness_new(self, nlags=10, signif=0.05, adjusted=False):
```

Comment:
@josef-pkt This looks like what we've discussed in #3683.  It's almost identical to the version I whipped up this morning.

@yogabonito Do you have a good intuition for why the `k_ar` shows up in the degrees of freedom?  The Stata docs I found for this were also VAR-specific, but it is not clear to me why.  It seems like this kind of test should be applicable to more general time-series residuals.